### PR TITLE
add external resource parsing

### LINKED ISSUE #3266: Add external resource definition in the CLI
As a developer I want to be able to define external resources in my okteto manifest, so I can use that information in the UI to have a full view of my application.

The idea is to have a new key at the root of okteto manifest which defines the external resource that will be shown in the UI.

The definition of this new section is:
```golang
type Manifest struct {
  ... 
  ExternalResources map[string]ExternalResourceInfo
}

type externalResourceInfoUnmarshall struct {
  notes string // path from the root of the project
  endpoints []ExternalEndpoints
}

type externalResourceInfo struct {
  notes ExternalResourceNotes
  endpoints []ExternalEndpoints
}

type ExternalResourceNotes struct {
  path string
  markdown string // base64 encoded content of the path
}

type externalResourceEndpoints struct {
  name string // endpoint name
  url string // URL to display in the UI/endpoints cmd
}
``` 

## Acceptance criteria

- [x] Define the external resource structs
- [x] It should be unmarshalled correctly
  - [x] notes must not be empty
  - [x] there must be at least one endpoint
- [ ] It should show a warning and skip it if the user is not in an okteto cluster. 
  - [ ] The warning must be shown only if the external key is present 
- [x] The environment variable `$OKTETO_EXTERNAL_($NAME)_ENDPOINTS_($NAME)_URL` must be available in the deploy section
- [x] Differentiate between the unmarshall struct and the one being used in the rest of the command
- [x] Store the content of `external.$NAME.notes` in base64

### PR DESCRIPTION
**Pull request by AdrianPedriza**
Signed-off-by: adripedriza <adripedriza@gmail.com>

# Proposed changes
- add reading of an external resource declaration in a manifest
- add method for creating required environment variables
- add method to load markdown content and store it in base64
- add test covering features

Fixes #3266

# Notes: 
the warning if not in an okteto cluster will be added once the integration with the commands is started.

### CODE REVIEW

File: pkg/externalresource/serializer.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
+package externalresource
+
+import (
+	"fmt"
+)
+
+type externalResourceRaw struct {
```
Comment (jLopezbarb) [inline pkg/externalresource/serializer.go:7]:
```suggestion
type externalResourceUnmarshaller struct {
```

File: pkg/externalresource/external_resource.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	b64 "encoding/base64"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/spf13/afero"
+)
+
+// ExternalResources represents the map of external resources at a manifest
+type ExternalResources map[string]*ExternalResource
```
Comment (jLopezbarb) [inline pkg/externalresource/external_resource.go:13]: Let's use a name that can differentiate more from the actual type

```suggestion
type ExternalResourceSection map[string]*ExternalResource
```

File: pkg/externalresource/external_resource.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+// ERFilesystemManager represents ExternalResource information with the filesystem inyected
+type ERFilesystemManager struct {
+	ExternalResource ExternalResource
+	Fs               afero.Fs
+}
+
+// SetDefaults creates the necessary environment variables given an external resource
+func (er *ExternalResource) SetDefaults(externalName string) {
+	for _, endpoint := range er.Endpoints {
+		endpointUrlEnv := fmt.Sprintf("OKTETO_EXTERNAL_%s_ENDPOINTS_%s_URL", externalName, endpoint.Name)
```
Comment (jLopezbarb) [inline pkg/externalresource/external_resource.go:43]: We would probably need to sanitize `endpoint.Name`. We need to uppercase the name and replace `-` with `_`

File: pkg/externalresource/external_resource.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+// ERFilesystemManager represents ExternalResource information with the filesystem inyected
+type ERFilesystemManager struct {
+	ExternalResource ExternalResource
+	Fs               afero.Fs
+}
+
+// SetDefaults creates the necessary environment variables given an external resource
+func (er *ExternalResource) SetDefaults(externalName string) {
```
Comment (AgustinRamiroDiaz) [inline pkg/externalresource/external_resource.go:42]: This is only used in the test. I don't think it should be exported, and maybe it should be in the test itself

File: pkg/externalresource/external_resource.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+		os.Setenv(endpointUrlEnv, endpoint.Url)
+	}
+}
+
+func sanitizeForEnv(name string) string {
+	return strings.ToUpper(strings.ReplaceAll(name, "-", "_"))
+}
+
+// LoadMarkdownContent loads and store markdown content related to external resource
+func (ef *ERFilesystemManager) LoadMarkdownContent(manifestPath string) error {
```
Comment (AgustinRamiroDiaz) [inline pkg/externalresource/external_resource.go:56]: This is only used in the test. Why do we have these functions?

Comment (AdrianPedriza) [reply pkg/externalresource/external_resource.go:56]: yes at the moment these functions are not used at any place but both will be used by the commands. It is due to the way the epic work has been split up

File: pkg/externalresource/external_resource.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+// ExternalEndpoint represents information about an endpoint
+type ExternalEndpoint struct {
+	Name string
+	Url  string
+}
+
+// ERFilesystemManager represents ExternalResource information with the filesystem inyected
+type ERFilesystemManager struct {
```
Comment (jLopezbarb) [inline pkg/externalresource/external_resource.go:36]: This is the only thing that annoys me... We create an struct to modify one of their fields. I think it can be better if we return externalresource modified. WDYT?

File: pkg/externalresource/external_resource.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+	b64 "encoding/base64"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/spf13/afero"
+)
+
+// ExternalResources represents the map of external resources at a manifest
```
Comment (teresaromero) [inline pkg/externalresource/external_resource.go:13]:
```suggestion
// ExternalResourcesSection represents the map of external resources at a manifest
```

File: pkg/externalresource/external_resource.go [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+func (er *ExternalResource) SetDefaults(externalName string) {
+	sanitizedExternalName := sanitizeForEnv(externalName)
+	for _, endpoint := range er.Endpoints {
+		sanitizedEndpointName := sanitizeForEnv(endpoint.Name)
+		endpointUrlEnv := fmt.Sprintf("OKTETO_EXTERNAL_%s_ENDPOINTS_%s_URL", sanitizedExternalName, sanitizedEndpointName)
+		os.Setenv(endpointUrlEnv, endpoint.Url)
+	}
+}
+
+func sanitizeForEnv(name string) string {
```
Comment (teresaromero) [inline pkg/externalresource/external_resource.go:51]: what about spaces and special characters? does this name need to be part of any kubernetes resource?

File: pkg/externalresource/serializer.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+	for _, entry := range result.Endpoints {
+		if _, isAdded := uniqueEndpointsNames[entry.Name]; isAdded {
+			return fmt.Errorf("there must be no duplicate names for the endpoints of an external resource")
+		}
+
+		uniqueEndpointsNames[entry.Name] = false
+	}
+
+	er.Icon = result.Icon
+	er.Notes.Path = result.Notes
```
Comment (teresaromero) [inline pkg/externalresource/serializer.go:43]: for the moment only the Notes.Path is unmarshalled here? could we add a note TODO for the Markdown prop of the Notes? or where this would be done?

File: pkg/externalresource/external_resource.go (hunk 5 above)
Comment (ifbyol) [reply pkg/externalresource/external_resource.go:56]: Shouldn't/can't be part of the manifest deserialization? I mean, if we are reading all the `ExternalResource` from the manifest definition, wouldn't it make sense to read the markdown there? If not, where would we be reading it?

File: pkg/externalresource/external_resource.go (hunk 8 above)
Comment (AdrianPedriza) [reply pkg/externalresource/external_resource.go:51]: this sanitization is to incorporate the name of the external and its endpoints as an environment variable. But yes, it would be necessary to avoid blank spaces

File: pkg/externalresource/external_resource.go (hunk 5 above)
Comment (AdrianPedriza) [reply pkg/externalresource/external_resource.go:56]: I need to know somehow the path of the manifest to build the path to the markdown and that inside the deserialization process I can't know it I think

File: pkg/externalresource/serializer.go (hunk 9 above)
Comment (AdrianPedriza) [reply pkg/externalresource/serializer.go:43]: in principle I cannot add the markdown content because I need to know the path to the manifest

File: pkg/externalresource/external_resource.go (hunk 6 above)
Comment (AdrianPedriza) [reply pkg/externalresource/external_resource.go:36]: In my opinion I think that the name of the function helps not to create confusion, it simply loads the content of the markdown. Receiving as parameter a struct and returning it modified or creating a new one I believe that leaves us in a similar situation and even you are left with a more verbose function declaration. But that's my opinion, I have no strong opinion for or against it

Comment (jLopezbarb) [reply pkg/externalresource/external_resource.go:36]: What if is `ExternalResource` the one that has the afero dependency? That way we don't have to create another an intermediate struct. `externalResource.LoadMarkdownContent(path string)`. 

We could even not add the afero dependency as a field and initialise it inside the `externalResource.LoadMarkdownContent`.

Comment (AdrianPedriza) [reply pkg/externalresource/external_resource.go:36]: in most of the use of this struct this property would not be used, I believe that it is something foreign to the information that must store the `ExternalResource` struct. If we instantiate it inside the function we can not use the injection of this variable for its testing

**Review by jLopezbarb — APPROVED** [review]

File: pkg/externalresource/external_resource.go (hunk 5 above)
Comment (ifbyol) [reply pkg/externalresource/external_resource.go:57]: oh, ok, I see. Makes sense then to have it outside the parsing

File: pkg/model/dev.go [hunk 10]
```diff
@@ -892,7 +892,7 @@ func (dev *Dev) LoadRemote(pubKeyPath string) {
 	p := Secret{
 		LocalPath:  pubKeyPath,
 		RemotePath: authorizedKeysPath,
-		Mode:       0600,
+		Mode:       0644,
```
Comment (ifbyol) [inline pkg/model/dev.go:895]: Why this change? Wasn't it already on master?

**Review by ifbyol — APPROVED** [review]
Just check the comment about the change related to permissions. I though that was already merged. Besides that, everything looks fine!

File: pkg/model/dev.go (hunk 10 above)
Comment (AdrianPedriza) [reply pkg/model/dev.go:895]: :hushed: some failure updating with master. I will check it. Thanks Nacho!
